client/app/private-database/user/grants/private-database-user-grants.controller.js

```diff
@@ -94,72 +89,31 @@ angular.module("App").controller(
             this.privateDatabaseService.restartPoll(this.productId, ["grant/create", "grant/update"]);
         }
 
-        /*
-         * Grant User jobs
-         */
-        onUserGrantstart (evt, opts) {
-            let unregister = null;
-
-            const todo = () => {
-                const idx = _.findIndex(this.$scope.userGrants, (grant) => grant.dataBase === opts.databaseName);
-
-                if (~idx) {
-                    this.isDoingGrant[opts.databaseName] = true;
-                    if (unregister) {
-                        unregister();
-                    }
-                }
-            };
-
-            if (this.$scope.userGrants && this.$scope.userGrants.length) {
-                todo();
-            } else {
-                unregister = this.$scope.$watch("userGrants.length", todo);
-            }
+        bindGrantStatusToMethods (statusToWatch) {
+            _.forEach(statusToWatch, (state) => {
+                this.$scope.$on(`privateDatabase.grant.set.${state}`, this[`onUserGrant_${state}`].bind(this));
+            });
         }
 
-        onUserGrantdone (evt, opts) {
-            let unregister = null;
-
-            const todo = () => {
-                this.isDoingGrant[opts.databaseName] = false;
-
-                this.userGrants[opts.databaseName].value = opts.grants[opts.databaseName].value;
-
-                this.alerter.success(this.$scope.tr("privateDatabase_tabs_users_grant_success"), this.$scope.alerts.main);
-
-                this.getUserGrants();
-
-                if (unregister) {
-                    unregister();
-                }
-            };
-
-            if (this.$scope.userGrants && this.$scope.userGrants.length) {
-                todo();
-            } else {
-                unregister = this.$scope.$watch("users.length", todo);
+        onUserGrant_start (evt, opts) {
+            if (_.any(this.userGrants, { dataBase: opts.databaseName })) {
```

Comment:
`this.isDoingGrant[opts.databaseName] = _.any(this.userGrants, { dataBase: opts.databaseName })`